### PR TITLE
[WIP] [PORT] Adds Lawnmowers For Janitors

### PR DESCRIPTION
**Pull request by VaelophisNyx**
# Document the changes in your pull request
Ports from Whitesands/Hippiestation

Adds lawnmowers to the game, which are keyed for janitors and upgraded similar to the janitor's cart
So far they can mow down any subtype of structure/flora (bushes, grass, flowers) and glowshrooms, and if upgraded, spacevines and alien weeds

Planned features:

- [ ] require it to contain fuel to run
- [x] make emagged mowers damage you, or delimb & decapitate if upgraded as well
- [x] make upgrading it expand the list of mowable structures
- [ ] get better sounds
- [ ] add a startup/shutoff noise
- [x] add it to cargo

# Spriting

New sprites needed, current sprite is a placeholder from Whitesands

- [ ] Mower Sprites (4 facings) (Janitorial colors)
- [ ] Overlay for upgraded mower deck ^
- [ ] Destroyed Mower Sprite
- [ ] Energy Blades Upgrade Sprite

# Wiki Documentation

E

# Changelog

:cl:  @Chubbygummibear, VaelophisNyx, @tattax 
rscadd: Adds Riding Lawnmowers to the game
/:cl:

### CODE REVIEW

**Comment by VaelophisNyx** [comment]
Opening this so people can discuss it, and so I can't forget as easily.

File: code/modules/vehicles/lawnmower.dm [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+	var/emagged = FALSE
+	var/fuel = 0
+	var/obj/item/reagent_containers/glass/mycontainer = null
+	var/static/list/mowable = typecacheof(list(/obj/structure/flora, /obj/structure/glowshroom))
+
+/obj/vehicle/ridden/lawnmower/Initialize(mapload)
+	. = ..()
+	mow_lawn()
+	mycontainer = list()
+	mycontainer += new /obj/item/reagent_containers/glass/beaker/large(null)
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
	mycontainer = list(new /obj/item/reagent_containers/glass/beaker/large(null), )
```

File: code/modules/vehicles/lawnmower.dm [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+/obj/vehicle/ridden/lawnmower/examine(mob/user)
+	. += ..()
+	if(bladeattached)
+		. += "It has been upgraded with a energy blade."
+
+/obj/vehicle/ridden/lawnmower/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/storage/bag/trash))
+		if(mybag)
+			to_chat(user, span_warning("[src] already has a trashbag hooked!"))
+			return
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
		if(mybag && !istype(I, /obj/item/storage/bag/trash/bluespace)
			to_chat(user, span_warning("[src] already has a trashbag hooked!"))
			return
		mybag.forceMove(get_turf(user))
```

Comment (tattax) [reply code/modules/vehicles/lawnmower.dm]: upgradable bag

**Comment by GraveHat** [comment]
Making a new subtype that only janitors and botanists can buy or giving the emagged version a way to sacrifice fuel for speed or a dash would be a nice addition to make it not completely horrible and maybe somewhat viable traitor option if you want it to be less of a joke item they can get.

File: code/modules/vehicles/lawnmower.dm [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+	if(mybag)
+		add_overlay("mower_bag")
+	if(bladeattached)
+		add_overlay("mower_blade")
+
+/obj/vehicle/ridden/lawnmower/attack_hand(mob/user)
+	. = ..()
+	if(.)
+		return
+	else if(mybag)
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
	if(mybag)
```

File: code/modules/vehicles/lawnmower.dm [hunk 4]
```diff
@@ -0,0 +1,115 @@
+/obj/vehicle/ridden/lawnmower
+	name = "lawnmower"
+	desc = "Somewhat of a relic of pre-spaceflight, still seen on colonies for its utility, and on stations for unusual emergencies. Has electronic safties to prevent accidents."
+	icon_state = "lawnmower"
+	key_type = /obj/item/key/janitor
+	are_legs_exposed = TRUE
+	var/obj/item/storage/bag/trash/mybag = null
+	var/bladeattached = FALSE
+	var/emagged = FALSE
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
```

File: code/modules/vehicles/lawnmower.dm [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+/obj/vehicle/ridden/lawnmower/driver_move(mob/user, direction)
+	. = ..()
+	mow_lawn()
+
+/obj/vehicle/ridden/lawnmower/upgraded
+	bladeattached = TRUE
+
+/obj/vehicle/ridden/lawnmower/emagged
+	emagged = TRUE
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
	obj_flags |= EMAGGED
```

File: code/modules/vehicles/lawnmower.dm [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+/obj/vehicle/ridden/lawnmower/upgraded/emagged
+	bladeattached = TRUE
+	emagged = TRUE
+
+/obj/vehicle/ridden/lawnmower/emag_act(mob/user)
+	if(emagged)
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
	if(obj_flags & EMAGGED)
```

File: code/modules/vehicles/lawnmower.dm [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+		to_chat(user, span_warning("The safety mechanisms on [src] are already disabled!"))
+		return
+	to_chat(user, span_warning("You disable the safety mechanisms on [src]."))
+	emagged = TRUE
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
	obj_flags |= EMAGGED
```

File: code/modules/vehicles/lawnmower.dm [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+/obj/vehicle/ridden/lawnmower/upgraded/emag_act(mob/user)
+	if(emagged)
+		to_chat(user, span_warning("The safety mechanisms on [src] are already disabled!"))
+		return
+	to_chat(user, span_warning("You disable the safety mechanisms on [src]."))
+	emagged = TRUE
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
```

File: code/modules/vehicles/lawnmower.dm [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+/obj/vehicle/ridden/lawnmower/proc/mow_lawn()
+	var/mowed = FALSE
+	for(var/obj/structure/S in loc)
+		if(is_type_in_typecache(S, mowable))
+			qdel(S)
+			mowed = TRUE
+		if(mowed)
+			playsound(loc, pick(list("sound/vehicles/mowermove1.ogg", "sound/vehicles/mowermove2.ogg")), 50, 1)
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
			playsound(loc, pick(list("sound/vehicles/mowermove1.ogg", "sound/vehicles/mowermove2.ogg")), 50, TRUE)
```

File: code/modules/vehicles/lawnmower.dm (hunk 2 above)
Comment (VaelophisNyx) [reply code/modules/vehicles/lawnmower.dm]: I actually axed this on my local code entirely; forgot to remove this reference. Originally it was going to produce stuff for mowing...but couldn't find a reason to keep it.

File: code/modules/vehicles/lawnmower.dm [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+	mycontainer = /obj/item/reagent_containers/glass/beaker/large //start with a large beaker so its usable for a decent time
+	for(var/reagent_id in consumed_reagents_list)
+		if(!reagents.has_reagent(reagent_id, consumed_reagents_list[reagent_id]))
+			hasfuel = TRUE
+	update_icon()
+	var/datum/component/riding/D = LoadComponent(/datum/component/riding)
+	D.set_riding_offsets(RIDING_OFFSET_ALL, list(TEXT_NORTH = list(0, 4), TEXT_SOUTH = list(0, 7), TEXT_EAST = list(-5, 2), TEXT_WEST = list( 5, 2)))
+
+/obj/vehicle/ridden/lawnmower/Destroy()
+	. = ..()
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm:38]:
```suggestion
```

File: code/modules/vehicles/lawnmower.dm [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+/obj/item/mowerupgrade
+	name = "energy blade upgrade"
+	desc = "An upgrade for lawnmowers, allowing them to cut through dense organic matter."
+	icon = 'icons/obj/vehicles.dmi'
+	icon_state = "upgrade"
+
+/obj/vehicle/ridden/lawnmower/examine(mob/user)
+	. += ..()
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm:47]:
```suggestion
	. = ..()
```

File: code/modules/vehicles/lawnmower.dm [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+		I.play_tool_sound(src, 50)
+		if(!panel_open)
+			panel_open = TRUE
+			to_chat(user, span_notice("You open the service panel of [src]."))
+		else
+			panel_open = FALSE
+			to_chat(user, span_notice("You close the service panel of [src]."))
+		return 1
+	return 0
+
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
		return TRUE
	return FALSE
```

File: code/modules/vehicles/lawnmower.dm [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+/obj/vehicle/ridden/lawnmower/proc/default_deconstruction_screwdriver(mob/user, obj/item/I)
+	if(I.tool_behaviour == TOOL_SCREWDRIVER)
+		I.play_tool_sound(src, 50)
+		if(!panel_open)
+			panel_open = TRUE
+			to_chat(user, span_notice("You open the service panel of [src]."))
+		else
+			panel_open = FALSE
+			to_chat(user, span_notice("You close the service panel of [src]."))
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
		panel_open = !panel_open
		to_chat(user, span_notice("You [panel_open ? "open" : "close"] the service panel of [src]."))
```

File: code/modules/vehicles/lawnmower.dm [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+	icon_state = "upgrade"
+
+/obj/vehicle/ridden/lawnmower/examine(mob/user)
+	. += ..()
+	if(bladeattached)
+		. += "It has been upgraded with energy blades."
+	if(panelopen)
+		. += "Its service panel is open!"
+	if(obj_flags & EMAGGED)
+		. += "the safety lights appear to be flickering"
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm:58]:
```suggestion
		. += "The safety lights appear to be flickering."
```

File: code/modules/vehicles/lawnmower.dm [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+		return TRUE
+
+	if (istype(I, /obj/item/reagent_containers/glass/) && !(I.item_flags & ABSTRACT) && I.is_open_container())
+		. = TRUE //no afterattack
+		reagents.maximum_volume += G.volume
+		if(!user.transferItemToLoc(G, src))
+			return
+		replace_container(user, G)
+		to_chat(user, span_notice("You add [G] to [src]."))
+		return TRUE //no afterattack
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
		if(replace_container(user, G))
			to_chat(user, span_notice("You add [G] to [src]."))
			return TRUE //no afterattack
		return
```

File: code/modules/vehicles/lawnmower.dm [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+
+/obj/vehicle/ridden/lawnmower/proc/UseFuel() //Turns out you need fuel to make this thing go!
+	if(hasfuel)
+
+		//world.time + fueluserate
+	
+/obj/vehicle/ridden/lawnmower/proc/ToggleEngine()
+	var/hasfuel = FALSE
+	if(G.total_reagents > 0)
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
	if(G.total_reagents)
```

File: code/modules/vehicles/lawnmower.dm [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+		to_chat(user, span_notice("You add [G] to [src]."))
+		return TRUE //no afterattack
+
+/obj/vehicle/ridden/lawnmower/proc/UseFuel() //Turns out you need fuel to make this thing go!
+	if(hasfuel)
+
+		//world.time + fueluserate
+	
+/obj/vehicle/ridden/lawnmower/proc/ToggleEngine()
+	var/hasfuel = FALSE
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
```

File: code/modules/vehicles/lawnmower.dm [hunk 18]
```diff
@@ -0,0 +1,166 @@
...
+
+/obj/vehicle/ridden/lawnmower/proc/FuelFill()
+
+/obj/vehicle/ridden/lawnmower/process()
+	if(active)
+		if(!hasfuel || !key_type)
+			ToggleEngine()
+			return
+		else
+			UseFuel()
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
		UseFuel()
```

File: code/modules/vehicles/lawnmower.dm [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+	else
+		return
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]:
```suggestion
	return
```

File: code/modules/vehicles/lawnmower.dm [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+	if(istype(I, /obj/item/mowerupgrade))
+		if(bladeattached)
+			to_chat(user, span_warning("[src] already has energy blades!"))
+			return
+		add_type_to_cache()
+		bladeattached = TRUE
+		qdel(I)
+		to_chat(user, span_notice("You upgrade [src] with energy blades."))
+	else
+		return ..()
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm]: move to fisrt proc

File: code/modules/vehicles/lawnmower.dm [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+/obj/vehicle/ridden/lawnmower/upgraded/emagged
+	bladeattached = TRUE
+	obj_flags = EMAGGED
+
+/obj/vehicle/ridden/lawnmower/emag_act(mob/user)
+	if(obj_flags & EMAGGED)
+		to_chat(user, span_warning("The safety mechanisms on [src] are already disabled!"))
+		return
+	to_chat(user, span_warning("You disable the safety mechanisms on [src]."))
+	obj_flags = EMAGGED
```
Comment (tattax) [inline code/modules/vehicles/lawnmower.dm:216]:
```suggestion
	obj_flags |= EMAGGED
```

**Comment by VaelophisNyx** [comment]
Hang on, been working with @Chubbygummibear a LOT in the background and most of this is either addressed or replaced entirely. I'll ping you for rereview when it's pushed

also dont forget the obj_flag |= wasn't working; but just = does

**Comment by VaelophisNyx** [comment]
I'll come back to this one at some point in the future. It's been a rollercoaster of shit working, then breaking

**Pull request closed without merging**